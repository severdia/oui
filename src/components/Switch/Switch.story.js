import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';

import Switch from './index.js';

const stories = storiesOf('Switch', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories.add('Default', (() => {
  return <Switch onClick={ action('clicked the switch') }/>;
}));

stories.add('Off', (() => {
  return <Switch checked={ false } onClick={ action('clicked the switch') }/>;
}));

stories.add('On', (() => {
  return <Switch checked={ true } onClick={ action('clicked the switch') }/>;
}));

stories.add('Disabled', (() => {
  return (
    <div>
      <Switch checked={ true } isDisabled={ true } onClick={ action('Checked') }/>
      <Switch checked={ false } isDisabled={ true } onClick={ action('Not checked') }/>
    </div>
  );
}));
