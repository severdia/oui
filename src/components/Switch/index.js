import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const Switch = (props) => {
  const {
    onClick,
    elementId = 'switch1',
    checked,
    isDisabled,
    testSection,
  } = props;

  const switchClasses = classnames({
    'oui-switch': true,
    'oui-switch--disabled': isDisabled,
  });

  return (
    <div
      className='position--relative'
      data-oui-component={ true }
      data-test-section={ testSection }>
      <input
        type="checkbox"
        id={ elementId }
        className={ switchClasses }
        checked={ checked }
        onClick={ onClick }
        disabled={ isDisabled }
      />
      <label
        htmlFor={ elementId }
        data-on-label="On"
        data-off-label="Off"></label>
    </div>
  );
};

Switch.propTypes = {
  // checked true means the switch is on; false is off
  checked: PropTypes.bool,
  // string for label/input id pair, should be unique to the page
  elementId: PropTypes.string,
  /** Whether it is disabled, will render greyscale if so **/
  isDisabled: PropTypes.bool,
  // onClick function
  onClick: PropTypes.func,
  /** Hook for automated JavaScript tests */
  testSection: PropTypes.string,
};

export default Switch;
