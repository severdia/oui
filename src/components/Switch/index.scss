////
/// Switch
/// @description switch a state to on/off.
///
/// @example[html] Default spinner
/// <input type="checkbox" id="switch1" class="oui-switch" />
/// <label for="switch1" data-on-label="On" data-off-label="Off"></label>

/* don't hide input from screen-readers and keyboard access */
.oui-switch {
  position: absolute;
  opacity: 0;
  z-index: 3;
}
.oui-switch:checked ~ span a {
  right: 0%;
}

.oui-switch + label {
  font-size: 1em;
  line-height: 1;
  width: 4.3rem;
  height: 2.3rem;
  background-color: map-fetch($switch, background-color, off);
  background-image: none;
  border-radius: 2rem;
  padding: 0.1666666667rem;
  cursor: pointer;
  display: inline-block;
  text-align: center;
  position: relative;
  -webkit-box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.2) inset;
          box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.2) inset;
  font-family: inherit;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  -webkit-user-select: none;
  border: 2px solid white;
}

/* Label */
.oui-switch + label:before {
  color: #888;
  content: attr(data-off-label);
  display: block;
  font-family: inherit;
  font-family: FontAwesome, inherit;
  font-weight: 900;
  font-size: 0.85rem;
  line-height: 1.74rem;
  position: absolute;
  right: 0.2166666667rem;
  margin: 0.2166666667rem;
  top: 0;
  text-align: center;
  min-width: 1.6666666667rem;
  overflow: hidden;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
}

/* Slider */
.oui-switch + label:after {
  content: '';
  position: absolute;
  left: 0.1666666667rem;
  background-color: #f7f7f7;
  -webkit-box-shadow: none;
          box-shadow: none;
  border-radius: 2rem;
  height: 1.6666666667rem;
  width: 1.6666666667rem;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
}

.oui-switch:checked + label {
  background-color: map-fetch($switch, background-color, on);
}

.oui-switch:checked + label:before {
  color: #fff;
  content: attr(data-on-label);
  right: auto;
  left: 0.2166666667rem;
}

.oui-switch:checked + label:after {
  left: 2.1666666667rem;
  background-color: #f7f7f7;
  -webkit-box-shadow: 1px 1px 10px 0 rgba(0, 0, 0, 0.3);
          box-shadow: 1px 1px 10px 0 rgba(0, 0, 0, 0.3);
}

.oui-switch:focus + label, .oui-switch:active + label {
  box-shadow:
    1px 1px 10px rgba(0, 0, 0, 0.2) inset,
    0 0 0 1px var(--light-blue-25) !important;
}

.oui-switch--disabled + label {
  opacity: 0.4;
  pointer-events: none;
}
.oui-switch--disabled:checked + label {
  opacity: 0.2;
}
