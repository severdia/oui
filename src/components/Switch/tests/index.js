import React from 'react';
import Switch from '../index';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('components/Switch', () => {

  it('should render an `Switch` with correct props', () => {
    const component = shallow(<Switch />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render an `Switch` with correct props', () => {
    const component = shallow(<Switch />);
    expect(component.length).toBe(1);
  });

  it('should disable the switch', () => {
    const component = mount(<Switch isDisabled={ true } />);
    expect(component.find('input').hasClass('oui-switch--disabled')).toBe(true);
  });

  it('should disable the input', () => {
    const component = mount(<Switch isDisabled={ true } />);
    expect(component.find('input').is('[disabled]')).toBe(true);
  });

});
