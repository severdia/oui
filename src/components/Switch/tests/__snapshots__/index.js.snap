// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Switch should render an \`Switch\` with correct props 1`] = `
<div
  className="position--relative"
  data-oui-component={true}
>
  <input
    className="oui-switch"
    id="switch1"
    type="checkbox"
  />
  <label
    data-off-label="Off"
    data-on-label="On"
    htmlFor="switch1"
  />
</div>
`;
